Add a specialised TLS 1.3 KDF implementation

paulidale opened the pull request:
This is HKDF underneath but there is some additional byte marshaling that needs to be done inside the FIPS boundary.
This is required so that a power up self test can be used to validate correct operation under the TLS assumptions.

- [x] documentation is added or updated
- [x] tests are added or updated

slontis commented on doc/man7/EVP_KDF-TLS13_KDF.pod:
```diff
@@ -0,0 +1,116 @@
+These parameters work as described in L<EVP_KDF(3)/PARAMETERS>.
+
+=item "prefix" (B<OSSL_KDF_PARAM_PREFIX>) <octet string>
+=item "label" (B<OSSL_KDF_PARAM_LABEL>) <octet string>
```
blankline between =item fields?

paulidale replied:
Forgot I'd not finished all of the documentation......
Fix coming.

paulidale replied:
And done.

slontis commented on doc/man7/EVP_KDF-TLS13_KDF.pod:
```diff
@@ -0,0 +1,116 @@
+
+=item "prefix" (B<OSSL_KDF_PARAM_PREFIX>) <octet string>
+=item "label" (B<OSSL_KDF_PARAM_LABEL>) <octet string>
+=item "data" (B<OSSL_KDF_PARAM_DATA>) <octet string>
```
There is no description for prefix, label and data

paulidale replied:
Rather, these need documenting :)

slontis commented on doc/man7/EVP_KDF-TLS13_KDF.pod:
```diff
@@ -0,0 +1,116 @@
+L<EVP_KDF_CTX_set_params(3)> must detail all parameters required.
+This KDF does not support a piecemeal approach to providing these.
+
+A context for HKDF can be obtained by calling:
```
HKDF?

paulidale replied:
Fixed.

slontis commented on providers/fips/self_test_data.inc:
```diff
@@ -494,8 +494,76 @@ static const ST_KAT_PARAM kbkdf_params[] = {
+static const unsigned char tls13_kdf_prefix[] = {
+    0x74, 0x6C, 0x73, 0x31, 0x33, 0x20          /* "tls13 " */
+};
+static const unsigned char tls13_kdf_client_early_secret_label[] = {
```
Did these values come from ACVP test data?

paulidale replied:
Mostly.  All values except the intermediate early secret are.  The intermediate value (`tls13_kdf_early_secret`) wasn't specified but it is the output of the first test and the input to the second, so everything is okay end to end.

mattcaswell replied:
A comment stating where they came from would be useful

paulidale replied:
I've added a comment.

I'll note that none of the other self test KAT data includes a source reference.

slontis commented on include/openssl/self_test.h:
```diff
@@ -68,6 +68,7 @@ extern "C" {
 # define OSSL_SELF_TEST_DESC_KDF_SSHKDF     "SSHKDF"
 # define OSSL_SELF_TEST_DESC_KDF_TLS12_PRF  "TLS12_PRF"
 # define OSSL_SELF_TEST_DESC_KDF_KBKDF      "KBKDF"
+# define OSSL_SELF_TEST_DESC_KDF_TLS13_KDF  "TLS13_KDF"
```
doc/man7/OSSL_PROVIDER-FIPS.pod also needs updating.

I would suggest using 2 names here "TLS13_KDF_EXTRACT" and ""TLS13_KDF_EXPAND" maybe? And then just using those names in the self test.

slontis replied:
Or PSK??

paulidale replied:
I went with extract and expand.  Otherwise, I'd need a second meaning ful name.

slontis commented on providers/implementations/kdfs/hkdf.c:
```diff
@@ -496,3 +530,219 @@ static int HKDF_Expand(const EVP_MD *evp_md,
+
+    OPENSSL_clear_free(ctx->data, ctx->data_len);
+    ctx->data = NULL;
+    ctx->data_len = 0;
```
Why do you clear this length but not the other 2 above?

paulidale replied:
Didn't copy/pase this chunk :)

paulidale commented:
Feedback addressed.

slontis reviewed (approved):
Took a while to get my head around this, but the changes look ok to me.

t8m commented:
This needs a formal OTC exception as this is clearly not just a bug fix.

t8m commented on doc/man7/EVP_KDF-TLS13_KDF.pod:
```diff
@@ -0,0 +1,128 @@
+support all the options and capabilities that HKDF does.
+
+The I<OSSL_PARAM> array passed to L<EVP_KDF_derive(3)> or
+L<EVP_KDF_CTX_set_params(3)> must detail all parameters required.
```
Maybe `must contain` ?

paulidale replied:
I've changed to _must specify all of the parameters required_.

t8m commented on doc/man7/EVP_KDF-TLS13_KDF.pod:
```diff
@@ -0,0 +1,128 @@
+
+=head1 DESCRIPTION
+
+Support for computing the TLS 1.3 flavour of the B<HKDF> KDF through
```
I probably would not call it a `flavour`. Maybe `TLS 1.3 instance of...`? Or `TLS 1.3 specific instance of...`

paulidale replied:
I'll change to: _TLS 1.3 version of the HKDF_

t8m replied:
I am not sure `version` is the right word either. It is not like the underlying HKDF is modified by TLS-1.3 in any means. What about `TLS 1.3 way of using the HKDF`?

t8m commented on doc/man7/provider-kdf.pod:
```diff
@@ -239,6 +239,21 @@ Sets the scrypt work factor parameter p in the associated KDF ctx.
 
+=item "prefix" (B<OSSL_KDF_PARAM_PREFIX>) <octet string>
+
+Sets the prefix string using by the TLS 1.3 flavour of HKDF in the
```
Again I wouldn't use `flavour` and similarly below.

t8m commented on providers/fips/self_test_data.inc:
```diff
@@ -494,8 +494,76 @@ static const ST_KAT_PARAM kbkdf_params[] = {
 
+#ifndef OPENSSL_NO_TLS1_3
+static const char tls13_kdf_digest[] = "SHA256";
+static int tls13_kdf_extract_mode = 1;
+static int tls13_kdf_expand_mode = 2;
```
Why the magic numbers and not the constants?

paulidale replied:
Good point.
Will change.

t8m commented on doc/man7/EVP_KDF-TLS13_KDF.pod:
```diff
@@ -0,0 +1,128 @@
+
+=over 4
+
+=item B<EVP_KDF_HKDF_MODE_EXTRACT_ONLY> "EXTRACT_ONLY"
```
It is quite confusing that the EVP_KDF_HKDF_MODE_EXTRACT_ONLY is the integer constant and not the "EXTRACT_ONLY" string. I see this was copied from the HKDF documentation, maybe it should be somehow fixed there as well?

paulidale replied:
Not sure what to do about these two...

t8m replied:
What about: `=item "EXTRACT_ONLY" or B<EVP_KDF_HKDF_MODE_EXTRACT_ONLY>` This makes the order of the values the same as of the possible types of the parameter above. And adding the `or` somewhat indicates that these two are actually a different thing.

paulidale replied:
That should work.

I was tempted to not support both types for this param in the TLS 1.3 KDF but sharing the param handling function sort of prevented this.

t8m commented on doc/man7/EVP_KDF-TLS13_KDF.pod:
```diff
@@ -0,0 +1,128 @@
+The digest, key and salt values must be set before a key is derived otherwise
+an error will occur.
+
+=item B<EVP_KDF_HKDF_MODE_EXPAND_ONLY> "EXPAND_ONLY"
```
Similarly as above.

t8m commented on providers/defltprov.c:
```diff
@@ -320,6 +320,10 @@ static const OSSL_ALGORITHM deflt_macs[] = {
     { PROV_NAMES_HKDF, "provider=default", ossl_kdf_hkdf_functions },
+#ifndef OPENSSL_NO_TLS1_3
+    { PROV_NAMES_TLS1_3_KDF, "provider=default",
+      ossl_kdf_tls1_3_prf_functions },
```
It is a little bit inconsistent that sometimes there is `tls1_3_prf` and elsewhere `TLS1_3_KDF`.

t8m commented on providers/implementations/kdfs/hkdf.c:
```diff
@@ -496,3 +534,218 @@ static int HKDF_Expand(const EVP_MD *evp_md,
+
+    ret = EVP_MD_get_size(md);
+    /* Ensure cast to size_t is safe */
+    if (!ossl_assert(ret >= 0))
```
Not sure this is internal-error only thing. I suppose if you have a MD provider that fails to report the size of the MD, you would get -1? Also allowing `ret == 0` does not make any sense.

paulidale replied:
This was existing.  I wasn't judging it's suitability :)

t8m commented on providers/implementations/kdfs/hkdf.c:
```diff
@@ -496,3 +534,218 @@ static int HKDF_Expand(const EVP_MD *evp_md,
+    return known_settable_ctx_params;
+}
+
+const OSSL_DISPATCH ossl_kdf_tls1_3_prf_functions[] = {
```
Again `prf`

t8m reviewed (changes requested):
Good work! Some consistency and other nits.

t8m commented:
One more nit. Could you please add just a few testcases of this KDF to the evp test data? I do not care if they are just arbitrary values or anything from NIST or RFC data.

paulidale commented:
This is extensively tested via the TLS tests.  And specifically tested by `tls13secretstest.c`.

I didn't think additional case were warranted.

t8m commented:
> This is extensively tested via the TLS tests. And specifically tested by `tls13secretstest.c`.
> 
> I didn't think additional case were warranted.

I understand this is well-tested elsewhere but having it also in the general evp test would be good for consistency.

There is one more thing to consider - is it really needed to have this KDF conditional on OPENSSL_NO_TLS_1_3? I mean, the original implementation in tls13_enc.c is not conditional on OPENSSL_NO_TLS_1_3 either and I'd actually expect the OPENSSL_NO_TLS_1_3 to affect only the libssl in some ways not the providers.

paulidale commented:
There isn't any NIST test data -- NIST's test cover both operations sequentially as per my [earlier comment](https://github.com/openssl/openssl/pull/16203#discussion_r681445623) which doesn't fit evp_test so well.  I could run some cases and generate the intermediate to include.

paulidale commented:
OTC vote in progress.

paulidale commented:
I've dropped the TLS 1.3 conditional compilation.  That makes sense.
I've addressed the feedback except the additional test cases.

Test cases tomorrow.

paulidale commented:
Added a couple of test cases.

t8m commented:
@slontis still ok after the changes?

slontis commented:
reapproved.

paulidale commented:
I just pushed a pile more test cases base on ACVP data.
Reapprovals required.

t8m reviewed (approved):
Still good!

kaduk commented on include/openssl/core_names.h:
```diff
@@ -190,6 +190,9 @@ extern "C" {
 #define OSSL_KDF_PARAM_PASSWORD     "pass"      /* octet string */
+#define OSSL_KDF_PARAM_PREFIX       "prefix"    /* octet string */
+#define OSSL_KDF_PARAM_LABEL        "label"     /* octet string */
+#define OSSL_KDF_PARAM_DATA         "data"      /* octet string */
```
I know this isn't the first time, but it still feels a little weird to use such short/common words as parameters that are specific to the TLS 1.3 HKDF...

paulidale replied:
I half agree.  I'm not too concerned about the strings actually used.  So long as they have some correspondence to what is being passed, I'm okay.

I guess I don't see a problem reusing these labels in other contexts, so long as they are documented each time.

kaduk commented on providers/implementations/kdfs/hkdf.c:
```diff
@@ -496,3 +527,216 @@ static int HKDF_Expand(const EVP_MD *evp_md,
+ * |datalen| (e.g. typically a hash of the handshake messages), derive a new
+ * secret |outlen| bytes long and store it in the location pointed to be |out|.
+ * The |data| value may be zero length. Any errors will be treated as fatal if
+ * |fatal| is set. Returns 1 on success  0 on failure.
```
Where is `fatal`?

kaduk replied:
Is there a punctuation missing before "0 on failure"?

paulidale replied:
My oopses.

kaduk commented on providers/implementations/kdfs/hkdf.c:
```diff
@@ -496,3 +527,216 @@ static int HKDF_Expand(const EVP_MD *evp_md,
+    size_t hkdflabellen;
+    /*
+     * 2 bytes for length of derived secret + 1 byte for length of combined
+     * prefix and label + bytes for the label itself + 1 byte length of hash
```
does "label itself" include the prefix?
Maybe we don't need the comment at this level of detail...

paulidale replied:
It's a really good comment, but you are correct it's not critical anymore.

kaduk replied:
It might actually go better above the WPACKET stanza, come to think of it...

paulidale replied:
Okay, I've put it there and slightly reworded the final sentence.

paulidale commented on doc/man7/EVP_KDF-TLS13_KDF.pod:
```diff
@@ -0,0 +1,128 @@
+
+=head1 NOTES
+
+This KDF is intended for use by the TLS 1.3 stack in libssl.  It does not
```
Done

paulidale commented on doc/man7/EVP_KDF-TLS13_KDF.pod:
```diff
@@ -0,0 +1,128 @@
+A context for a TLS 1.3 KDF can be obtained by calling:
+
+ EVP_KDF *kdf = EVP_KDF_fetch(NULL, "TLS13-KDF", NULL);
+ EVP_KDF_CTX *kctx = EVP_KDF_CTX_new(kdf);
```
We don't in the other similar examples through the documentation.  I'm not going to add it just for this one -- it's special, it's internal and it's not for general consumption.  Adding it more globally would be a different PR I think.

mattcaswell reviewed (approved):
Nice work! This looks good to me. Approved whether or not you resolve the minor issue about adding a comment

mattcaswell reviewed (approved):
Nice work! This looks good to me. Approved whether or not you resolve the minor issue about adding a comment

paulidale commented:
There could also be a slight boost from not allocating the KDF twice for secret derivation :)

slontis commented:
Reapproved. Good work..

paulidale commented:
OTC vote passed.   Changes merged, thanks for the feedback.
